### PR TITLE
Additional sensors and binary sensors support for Haier Climate

### PR DESCRIPTION
**Pull request by paveldn**
# What does this implement/fix?

New sensors and binary sensors support for hOn protocol of Haier climate. Removed dependency on the sensor for climate. Fixed some warnings for hOn protocol if presets is not supported.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3628

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

uart:
  id: haier_uart
  baud_rate: 9600
  tx_pin: 1
  rx_pin: 3

climate:
  - platform: haier
    protocol: hOn
    id: haier_climate
    uart_id: haier_uart
    name: Haier AC

sensor:
  - platform: haier
    haier_id: haier_climate
    outdoor_temperature:
      name: Haier AC outdoor temperature
    humidity:
      name: Haier AC indoor humidity

binary_sensor:
  - platform: haier
    haier_id: haier_climate
    defrost_status:
      name: Haier AC defrost status

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/haier/binary_sensor/__init__.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    }
+).extend({cv.Optional(type): schema for type, schema in SENSOR_TYPES.items()})
+
+
+async def to_code(config):
+    paren = await cg.get_variable(config[CONF_HAIER_ID])
+
+    for type, _ in SENSOR_TYPES.items():
+        if type in config:
+            conf = config[type]
```
Comment (jesserockz) [inline esphome/components/haier/binary_sensor/__init__.py]:
```suggestion
        if conf := config.get(type):
```

File: esphome/components/haier/hon_packet.h [hunk 2]
```diff
@@ -105,40 +105,70 @@ struct HaierPacketControl {
...
+  // 40
+  uint8_t compressor_status : 2;               // 0 - off, 1 - on,  2 - information not available
+  uint8_t indoor_fan_status : 2;               // 0 - off, 1 - on,  2 - information not available
+  uint8_t four_way_valve_status : 2;           // 0 - off, 1 - on,  2 - information not available
+  uint8_t indoor_electric_heating_status : 2;  // 0 - off, 1 - on,  2 - information not available
+  // 41
+  uint8_t expansion_valve_open_degree[2];  // 0 - 4095
+};
+
+#define UINT16_BE(x) (((uint16_t) ((x)[0]) << 8) + (x)[1])
```
Comment (jesserockz) [inline esphome/components/haier/hon_packet.h]: We have `encode_uint16()` for this already
https://github.com/esphome/esphome/blob/f53f91e191e110c22aaa16c6f94f75d3d5ab6da7/esphome/core/helpers.h#L182-L184

File: esphome/components/haier/sensor/__init__.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    }
+).extend({cv.Optional(type): schema for type, schema in SENSOR_TYPES.items()})
+
+
+async def to_code(config):
+    paren = await cg.get_variable(config[CONF_HAIER_ID])
+
+    for type, _ in SENSOR_TYPES.items():
+        if type in config:
+            conf = config[type]
```
Comment (jesserockz) [inline esphome/components/haier/sensor/__init__.py]: Same as above
```suggestion
        if conf := config.get(type):
```

File: tests/test3.yaml
Comment (jesserockz) [inline tests/test3.yaml:1]: Could you please create test files for haier in `tests/components/haier`?
Take a look at https://github.com/esphome/esphome/pull/6179/files for the tests that were going to be added for haier already. You can completly copy/paste them into this PR no problem and make the required changes.

You can then revert the changes to this file for this PR.

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/haier/hon_packet.h (hunk 2 above)
Comment (paveldn) [reply esphome/components/haier/hon_packet.h]: Thanks, haven't seen this one before. Updated the code.

File: tests/test3.yaml
Comment (paveldn) [reply tests/test3.yaml:1]: Thanks, updated the tests and included them in this PR.

**Review by jesserockz — APPROVED** [review]
Thanks for making those changes

**Pull request merged**